x/dualstaking/keeper/delegate.go

```diff
@@ -0,0 +1,572 @@
+			types.ErrDelegationNotFound,
+			utils.Attribute{Key: "delegator", Value: delegator},
+			utils.Attribute{Key: "provider", Value: provider},
+		)
+	}
+
+	// if delegation now becomes zero, then remove this provider from the delegator
+	// entry; and if the delegator entry becomes entry then remove it altogether.
+	// otherwise just append the new version (for next epoch).
+	if delegationEntry.Amount.IsZero() {
+		delegatorEntry.DelProvider(provider)
+		if delegatorEntry.IsEmpty() {
+			err := k.delegatorFS.DelEntry(ctx, prefix, nextEpoch)
+			if err != nil {
+				// delete should never fail here
+				return utils.LavaFormatError("critical: delete delegator entry", err,
+					utils.Attribute{Key: "delegator", Value: delegator},
+					utils.Attribute{Key: "provider", Value: provider},
+				)
+			}
+		}
+	} else {
+		delegatorEntry.AddProvider(provider)
+		err := k.delegatorFS.AppendEntry(ctx, prefix, nextEpoch, &delegatorEntry)
+		if err != nil {
+			// append should never fail here
+			return utils.LavaFormatError("failed to update delegator entry", err,
+				utils.Attribute{Key: "delegator", Value: delegator},
+				utils.Attribute{Key: "provider", Value: provider},
+			)
+		}
+	}
+
+	// call updateStakeEntry() with the negative amount
+	amount = sdk.NewCoin(amount.Denom, sdk.ZeroInt()).Sub(amount)
+	if err := k.updateStakeEntry(ctx, provider, chainID, amount); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// updateStakeEntry updates the (epochstorage) stake-entry of the provider for the
+// given chain. amount can be positive or negative (to increase or decrease).
+func (k Keeper) updateStakeEntry(ctx sdk.Context, provider, chainID string, amount sdk.Coin) error {
+	providerAddr, err := sdk.AccAddressFromBech32(provider)
+	if err != nil {
+		// panic:ok: this call was alreadys successful by the caller
+		utils.LavaFormatPanic("updateStakeEntry: invalid provider address", err,
+			utils.Attribute{Key: "provider", Value: provider},
+		)
+	}
+
+	stakeEntry, exists, index := k.epochstorageKeeper.GetStakeEntryByAddressCurrent(ctx, chainID, providerAddr)
+	if !exists {
+		// if the stake entry is not found when we try to remove delegated funds,
+		// it means that the provider has been unstaked already: nothing to do.
+		if amount.IsNegative() {
+			return nil
+		}
+		// otherwise, we return an error because delegation to a provider that is
+		// not stake does not make sense.
+		return types.ErrProviderNotStaked
+	}
+
+	// sanity check
+	if stakeEntry.Address != provider {
+		return utils.LavaFormatError("critical: delagate to provider with address mismatch", sdkerrors.ErrInvalidAddress,
+			utils.Attribute{Key: "provider", Value: provider},
+			utils.Attribute{Key: "address", Value: stakeEntry.Address},
+		)
+	}
+
+	stakeEntry.DelegateTotal = stakeEntry.DelegateTotal.Add(amount)
+
+	// fail if amount was negative and the result underflowed the existing total
+	if stakeEntry.DelegateTotal.IsNegative() {
```

Comment:
if the amount is not negative than adding it to the total will always be positive

Reply:
this function may be called with positive amount or negative amount (e.g. redelegate,unbond)
in case of the latter, this test checks that the amount was not too big to have causes negative outcome - which would mean there were not enough funds deletegated to the provider to begin with.